### ISSUE TITLE
Alternative builder strategy

### ISSUE THREAD

**Issue by ibaca** [hal/elemento#20]
(Iuff... I'll update this comment later)

- AttacheableElement alternative: usually you either use inside-widget events or logic events, inside-widget event are safe to just leave it connected forever, ignoring the registration handler, otherwise, logic event frequently requires a correct event subscription handling, but as this is usually done in the presenter, this not need to be supported in the Builder. So, it is safe to just ignore the registration handlers in the builder.
- element stack might be useful but, as the most important point is type safety, this make much more difficult to use element type-specific builder, also the references and event handlers are only required because this stacks, if the stack is removed and a per element builder is created both implementation and use gets easier, and a builder-specific-per-type can be created.
- The current implementation uses inheritance to extends HTMLElement utilities, but there is an alternative that is using inheritance to create a more specific builder per element type.

https://github.com/ibaca/elemento
https://github.com/ibaca/rxtodo-gwt

This also revealed various problems with the old testing strategy, I have fixed using gwtmokito and making a dynamic mock of elements, but I'm not sure this is practical in the long term.

**Comment by ibaca** [comment]
I have made big changes in your initial builder. The most important are removing the stack strategy. This is common in other builders especially in js, BUT I think this is fine if you don't care too much for the specific type of each node. So although this looks nice at first...
```java
Elements.Builder builder = new Elements.Builder()
.li().css(item.completed ? "completed" : "").data("item", item.id)
    .div().css("view")
        .input(checkbox).on(change, event -> toggle()).rememberAs("toggle").css("toggle")
        .label().on(dblclick, event -> edit()).textContent(item.text).rememberAs("label").end()
        .button().on(click, event -> destroy()).css("destroy").end()
    .end()
    .input(text).on(keydown, this::keyDown).on(blur, event -> blur()).css("edit").rememberAs("input")
.end();

this.element = builder.buildAttachable();
this.li = element.asElement();
this.toggle = builder.referenceFor("toggle");
this.toggle.checked = item.completed;
this.label = builder.referenceFor("label");
this.input = builder.referenceFor("input");
```
This has varius problems. The most important, you don't have "type safety", either builder.build() or builder.referenceFor() just cast to any type, this is actually imposible (I think) to solve with the stack strategy. And "ironically" you need to use reference just because you use stacks. Both this problems (an one more described later) get solved with the builder per element strategy.

So my proposal is that each builder is a type safety element builder (similar to type safe events), and each builder just wraps an elemental2 element and exposes some really useful and common used method for this kind of elements. AND! ideally, this can be used like your already stack builder and define a view pretty similar as you do in ui.xml but programmatically and type-safe. This is how it looks with the per-element builder.
```java
this.container = li().data("item", item.id)
        .add(div().css("view")
                .add(toggle = input(checkbox).on(change, ev -> toggle()).css("toggle").asElement())
                .add(label = h("label").on(dblclick, ev -> edit()).textContent(item.text).asElement())
                .add(button().on(click, ev -> destroy()).css("destroy")))
        .add(summary = input(text).on(keydown, this::keyDown).on(blur, ev -> blur()).css("edit").asElement())
        .asElement();
this.container.classList.toggle("completed", item.completed);
this.toggle.checked = item.completed;
```
This is exactly the same, as you can see, the stack strategy actually end up adding more complexity both for the implementation and usability. This strategy forces the user to create each element independetly, and the "fieldName = <fieldBuilder>" to simulte the ui.xml ui:field="fieldName" assign. This in-builder field assignement might look ugly at first but it have the enormous advantages of been super simple and keep the View fields final! (which IMO is extremly useful). And, this can not be appreciated in this code, but now all this field has the specific type (and not using a cast to anything) of each element like HTMLButtonElement. Which, by the way, I thougth that those specific type do not add much, but it actually has a lot of fields! not sure this is a good idea to be used, maybe are not supported by all browsers, anyways, "real" typesafe is IMO important.

Finally, the inheritance situation. You are using inheritance to add "predefined element building" to a builder (at least in the example), and IMO it might be more useful if we use the inheritance to create a more element specific builder, for example there are some element that cannot contains other elements so an EmptyContentBuilder can be used  for those elements, others can only contain text, and there are a lot that can contain "anything" so TextContent and HtmlContent can be used by each one, etc.

The type-safety is done by a class pretty similar to EventType called ElementType<E,B>. E is the Element type and B is the Builder type. 

I have spent 1/3 of time just making test works again. You have a trick to make it work, but it only works because you don't use the concrete element type, I have enforced the specific type so creating a custom implementation of all the elements types is a nightmare. I have used gwtmockito which solves the native and final methods mocking, and I have created a dynamic element mocking. You can see it here (https://github.com/ibaca/elemento/commit/096d54a2f185caaea64dd85f4758a969d6c54d4a#diff-62daf40768eddcdb71dd6d4e42439559L28).

**Comment by ibaca** [comment]
This might not match your initial idea in this lib, so I totally understand you prefer to keep your initial strategy. I personally need to be able to build UI similar to ui.xml but programmatically and type-safe. But I really think this solves your problem, it is just a different strategy but can be used in a pretty similar way.

**Comment by hpehl** [comment]
Your approach looks very interesting. What like is the fact that it's type-safe and doesn't need castings. What I don't like is the additional `add()` and `asElement()` methods which makes it not as "fluent" as the stack solution. 

I'll take a closer look...

**Comment by ibaca** [comment]
Yep, I really try to maintain the fluent/stack strategy, but at the end, the per-element is much simple to implement and understand, more nice things about not-fluent:
- IDE indentation works again!
- don't need to remember or even now if you need to call end(), and of course, all internal validation to see if the stack is at 0 don't need to be done because the user is forced to do it right!
🥇 😜

**Comment by hpehl** [comment]
After looking at the code and working with the RX sample I really like your approach! The more I think about it, the more it makes sense. Once you get used to the inline field assignment it looks very similar to the stacked builder. 

Just a few questions / suggestions:

- What is the reason for having both an interfaces e.g. `HtmlContent` *and* an implementations e.g. `HtmlContentBuilder`? I would like to get rid of the interfaces and just use the typed builder implementations.
- Refactoring / Naming: I'd like to refactor / rename some builders to reflect the use case:
    - `EmptyContent` &larr; `SimpleTag`
    - Combine `TextContent` and `HtmlContent` into `BlockElement`

**Comment by ibaca** [comment]
HtmlContent (aka HtmlMixin), TextContent (aka TextMixin), and other builders should be an interface so it can be used as a mixin, this allows to do this `class MyBuilder extends BaseBuilder implements FooMixin, BarMixin`. In practice, any builder feature should be implemented as an interface, and you only use the actual class to make the builder `that()` works. Otherwise, you need to be sure and create a one-paren hierarchy which might not match js elements. But I'm not totally sure if it worth. We need to see more use cases (so more elements, ideally adding some web component or similar modern/advanced element) to actually decide if it worth it or if a simple hierarchy is enough.

I use the term `EmptyContent` because of this (https://developer.mozilla.org/en-US/docs/Web/HTML/Element/area)
![image](https://cloud.githubusercontent.com/assets/1042225/26559445/fb300178-44af-11e7-9a2b-3a86c9736694.png)
Are you sure that a user will understand that `Simple` means `An empty element is an element from HTML, SVG, or MathML that cannot have any child nodes (i.e., nested elements or text nodes).`. But if you still think it is better Simple I'm ok! I renamed various time class because I really don't like this name (EmptyContent) too, but it describes exactly the kind of element it builds.

Again, I'm ok with simplicity, so combining it (Text+Html) might be a good idea and I'm ok with it, but... are you sure you want to discard the "advantage" added with this independent mixins. You can see that there are any type of elements (https://developer.mozilla.org/en-US/docs/Web/HTML/Element), empty (no-content) elements, text-only elements, html elements, li-only elements, flow content elements, phrasing content events, etc.

Just to keep focus; this supposes to be a thin layer over elemental2 that should limit what the user can do with each type of element because JS is pretty flexible and exposes everything even if it is useless. So at the end, I think that the builder should allow discovering and use-without-runtime-surprises each element.

I say that because if an element can contain HTML but has no text content it can use the HTML mixin and not the text mixin (hehehe maybe this is impossible, not sure 😛). Anyway the mixin strategy is going to be useful, maybe you are right and all HtmlContent are TextContent, anyway using class instead of interfaces will force you yo create this hierarchy Base>Text>Html which it sounds to me that it's might not fit somewhere but right now I don't find any example.

**Comment by hpehl** [comment]
Makes sense to me. I've followed your naming scheme and merged your proposal into a feature branch: https://github.com/hal/elemento/tree/feature/typed-builder

If you're happy with it, I'll merge it into develop & master and release 0.4.0

**Comment by ibaca** [comment]
Oh, I see, you have moved inner classes out. Those are like internal or extension utilities, I suppose that this your main motivation to avoid them, and yes... all those classes there make it a bit confusing to the end user. Also, the 'dummy' classes should really be hidden inside some other class, those classes are only there so each element can be defined in a 120characters line (the generics were too large!). Also, the test commit is not correctly split. I'm going to update it correctly following your changes and splitting it in the same commits. Also making static fields private is a pretty good idea, actually just a few hours ago I was thinking that all those fields might be a problem, and I was thinking on grouping it in inner classes so it can be lazy static initialized, anyways making them all private make it easy to take this decision later.

Give me some day to try it out with some other internal project so I make sure all this builder makes sense or simplify them if it doesn't worth as you suggested. I don't go to touch your branch if you have suggestion put it there, and if you end up merging my branch will be nice 😉.

**Comment by ibaca** [comment]
After field privatization and other changes you made I see that the ElementType is not required anymore, this simplifies and reduces the number of classes, also the dummy types to reduce generics have been changed by normal method (no one line methods! also, removing ElementType make the generic simpler). Finally, I have moved builder related utils inside the builder package. I use inner classes before with the same purpose, users should not see all these classes in the core package or it will get too difficult to discover lib functionality just seen the package content. I personally like those libs where you see the main classes just there in the core package, and you can discover more advanced feature navigating into the classes or into some packages, but if you put all this builder util interfaces there it gets really messy.

**Comment by hpehl** [comment]
Where can I find your modifications?

**Comment by ibaca** [comment]
https://github.com/hal/elemento/commits/feature/typed-builder

**Comment by ibaca** [comment]
I initially force pushed this branch to change the author 😬, but later I see some "interesting" changes and I added on top of that. I hope you don't have any local change since your last commit.

**Comment by hpehl** [comment]
No problem. I found your commits and are about to merge with my local changes. I like the new builder package! 

Now some final cleanup, documentation and 0.4.0 is almost done.

**Comment by hpehl** [comment]
Released [0.4.0](https://github.com/hal/elemento/releases/tag/0.4.0)
